docs/en/platform-engineers/advanced-cue.md

```diff
@@ -0,0 +1,318 @@
+# Using CUE to Extend Trait in advanced way
+
+> WARNINIG: you are now reading a platform builder/administrator oriented documentation.
+
+In the following tutorial, you will learn how to add a trait in a more advanced way without writing any CRD controller.
+In general, the advanced way can help you build abstraction by composition or decomposition.
+
+## Trait generate multiple resources
+
+With the help of CUE template, we can combine multiple K8s resources into one trait.
+ 
+You can use the keyword `outputs` to create multiple K8s objects. The format MUST be `outputs:<unique-name>:<k8s-object>`.
+
+Let's look at an example, assume you hope to make a combo for K8s service and ingress, naming it as `new-route`.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: new-route
+spec:
+  extension:
+    template: |
+      parameter: {
+        domain: string
+        http: [string]: int
+      }
+
+      // trait template can have multiple outputs in one trait
+      outputs: service: {
+        apiVersion: "v1"
+        kind: "Service"
+        spec: {
+          selector:
+            app: context.name
+          ports: [
+            for k, v in parameter.http {
+              port: v
+              targetPort: v
+            }
+          ]
+        }
+      }
+
+      outputs: ingress: {
+        apiVersion: "networking.k8s.io/v1beta1"
+        kind: "Ingress"
+        metadata:
+          name: context.name
+        spec: {
+          rules: [{
+            host: parameter.domain
+            http: {
+              paths: [
+                for k, v in parameter.http {
+                  path: k
+                  backend: {
+                    serviceName: context.name
+                    servicePort: v
+                  }
+                }
+              ]
+            }
+          }]
+        }
+      }
+```
+
+Apply this newly defined TraitDefinition into our system:
+
+```shell script
+kubectl apply -f https://raw.githubusercontent.com/oam-dev/kubevela/master/docs/examples/advanced-cue/combo.yaml
+```
+
+Then the usage in is as usual in appfile, you need to use `vela traits` to refresh local capabilities and then use `vela up` to run.
+
+Here we use our server-side API `Application Object` for convenience, you don't rely on `vela` command line tool for this. 
```

Comment:
Don't do this.

There's no Appfile or vela CLI/dashboard in platform team doc. There's only k8s resources and kubectl.

Similarly, there's no k8s resources and kubectl in the end user doc, only Appfile and Vela interfaces make sense there.

Reply:
deleted